api/src/main/java/jakarta/data/repository/KeysetPageable.java

```diff
@@ -0,0 +1,265 @@
+/*
+ * Copyright (c) 2022 Contributors to the Eclipse Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ */
+package jakarta.data.repository;
+
+import java.util.Objects;
+
+/**
+ * <p>Keyset pagination is a form of pagination that aims to reduce the
+ * possibility of missed or duplicate results by making the request for
+ * each subsequent page relative to the observed values of entity properties
+ * from the current page. This list of values is referred to as the keyset
+ * and only includes values of entity properties that are in the sort criteria
+ * of the repository method. The values can be from the last entity (for
+ * forward direction) or first entity (for reverse direction) on the page,
+ * or can be any other desired list of values which serve as a new starting
+ * point.</p>
+ *
+ * <p>To use keyset pagination, define a repository method with return value of
+ * {@link KeysetAwareSlice} or {@link KeysetAwarePage} and which accepts a
+ * special parameter (after the normal query parameters) that is a
+ * {@link Pageable}. For example,</p>
+ *
+ * <pre>
+ * &#64;OrderBy("lastName")
+ * &#64;OrderBy("firstName")
+ * &#64;OrderBy("id")
```

Comment:
Not sure where to put that comment, but it might be worth pointing out that the fields by which the user orders by, should form a unique tuple for keyset pagination to work properly. Otherwise the database needs a notion of "ties" like in SQL the `... rows only with ties` clause to allow the server to fetch more rows, in case there are more with the same ordering.

Consider this:

```
users = [
{name: "A", id: 1},
{name: "B", id: 2},
{name: "B", id: 3},
{name: "C", id: 4}
]
```

When using `Pageable.size(2)` with just `@OrderBy("name")`, we will get either:

```
users = [
{name: "A", id: 1},
{name: "B", id: 2}
]
```

or:

```
users = [
{name: "A", id: 1},
{name: "B", id: 3}
]
```

and when paging to the next page:

```
users = [
{name: "C", id: 4}
]
```

because the data store uses `>`.

In Blaze-Persistence I validate the order by expressions and fail constructing a pagination query if I figure out during analysis that the expressions do not form a unique tuple.